drivers/ethernet/eth_stm32_hal.c

```diff
@@ -228,6 +229,61 @@ static void rx_thread(void *arg1, void *unused1, void *unused2)
 	}
 }
 
+#if defined(CONFIG_ETH_STM32_HAL_PHY_THREAD)
+static void phy_thread(void *arg1, void *unused1, void *unused2)
+{
+	struct device *dev;
+	struct eth_stm32_hal_dev_data *dev_data;
+	s64_t time_stamp;
+	u32_t phyreg;
+	int link_status;
+
+	__ASSERT_NO_MSG(arg1 != NULL);
+	ARG_UNUSED(unused1);
+	ARG_UNUSED(unused2);
+
+	dev = (struct device *)arg1;
+	dev_data = DEV_DATA(dev);
+
+	__ASSERT_NO_MSG(dev_data != NULL);
+
+	link_status = 0;
+
+	while (1) {
+
+		/* Get tick */
+		time_stamp = k_uptime_get();
+	    /* We wait for linked status */
+	    do
+	    {
+	        if((HAL_ETH_ReadPHYRegister(&dev_data->heth, MII_BMSR, &phyreg)) != HAL_OK) {
+	        	k_sleep(5);
+	        	continue;
+	        }
+
+			if(link_status != (phyreg & MII_BMSR_LINK_STATUS)) {
+
+				link_status = (phyreg & MII_BMSR_LINK_STATUS);
+
+				if(link_status) {
+				  /* link up event */
+					HAL_ETH_CheckAutoNegotiation(&dev_data->heth);
```

Comment:
What actions do you actually need from this new function?

Reply:
the HAL_ETH_CheckAutoNegotiation(..) function is the second part of the original HAL_ETH_Init(..) function which checks the actual PHY status and configure PHY, MAC and internal bookkeeping accordingly. 
So, the proposed solution is splitting the _init function in 2 parts. When the PHY thread is not used to actively check link status, the original code-progress is executed by first calling _init(..) and then _CheckAutoNegotiation(..). 
I added this split to touch the existing code as less as possible but still end up at the proposed end-situation.

Reply:
> checks the actual PHY status and configure PHY, MAC and internal bookkeeping accordingly

So, what is exactly required at this point and can't it be done thanks to existing functions like HAL_ETH_Read/WritePHYRegister, HAL_ETH_ConfigMAC, ... ?